### ISSUE TITLE
Magento\Framework\Model\AbstractModel::validateBeforeSave error notice (Array to string conversion)

### ISSUE THREAD

**Issue by stephan-cream** [magento/magento2#10207]
The function validateBeforeSave() on the Magento\Framework\Model\AbstractModel gives an error notice when fired "Notice: Array to string conversion in /vendor/magento/framework/Model/AbstractModel.php on line 694" while having a multi-dimensional array for messages.

### Preconditions
1. Any version of Magento 2 up to 2.3.2

### Steps to reproduce
1. Programmatically save a quote with an invalid email address.
2. Catch the error, try another save of the quote with an still invalid email address.

### Expected result
The correct validation errors are shown individually and no duplicates are shown.

### Actual result
An error notice is given "Notice: Array to string conversion in /vendor/magento/framework/Model/AbstractModel.php on line 694"

### What happens
$errors = $validator->getMessages() is an array, but in the implode() function it is expected that it is a simple array, however since the messages are added through array_merge_recursive() two messages for key 'invalid_email_format' are added making it a multi-dimensional array.

So its:
- Array ( [invalid_email_format] => Array ( [0] => Invalid email format [1] => Invalid email format) )
instead of the expected:
- Array ([invalid_email_format] => Invalid email format)

### Suggested fix
- Check if the $errors array is multi-dimensional, if so loop it to retreive the individual messages, merge out duplicates and add it to an new simple array.
- Use the newly made simple array for imploding the errors.

**Comment by magento-engcom-team** [comment]
@stephan-cream, thank you for your report.
We've created internal ticket(s) MAGETWO-81118 to track progress on the issue.

**Comment by valentinboyanov** [comment]
I am working on it at #mm17es

**Comment by rafaesteller** [comment]
I am working on it at #mm17es

**Comment by magento-engcom-team** [comment]
@stephan-cream, thank you for your report.     
    
Unfortunately, we are archiving this ticket now as it did not get much attention from both Magento Community and Core developers for an extended period. This is done in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to comment, reopen or create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
if you are still facing this issue on the latest `2.4-develop` branch. Thank you for collaboration.

**Comment by dverkade** [comment]
@magento-engcom-team You can't just close reproduced and validated issues!

**Comment by engcom-Charlie** [comment]
Hello @dverkade 
I can't reproduce this issue on the latest Magento 2.4-develop
Used customer data:
![info](https://user-images.githubusercontent.com/51680850/70641285-14c58980-1c46-11ea-884d-22d12fa335da.png)
Code(not all):
![quote](https://user-images.githubusercontent.com/51680850/70641315-1ee78800-1c46-11ea-934e-db30baa3562d.png)
**Result:**
![test](https://user-images.githubusercontent.com/51680850/70641355-2ad34a00-1c46-11ea-9c80-c0c2f3878cc5.png)

**Comment by stephan-cream** [comment]
**Preconditions**
Any version of Magento 2 up to 2.3.2 (untested on 2.3.3 and above)

**Steps to reproduce**
- Programmatically save a quote with an invalid email address.
- Catch the error, try another save of the quote with a still invalid email address.

In other words:

`try { // First attempt
    $this->quoteRepository->save($quote);
} catch (\Exception $exception) {
    try { // Second attempt
        $this->quoteRepository->save($quote);
    } catch (\Exception $exception) {
       $this->logger->error($exception->getMessage());
    }
}`

Basically save again within the `catch(...) {}` part of the code.

Hopefully this clarifies how to test this.

**Comment by engcom-Charlie** [comment]
Hello @stephan-cream 
This issue is reproducible on the Magento 2.3-develop(closed), but not on the 2.4-develop
I still can't reproduce this issue with your update:

```
<?php

namespace Vendor\Module\Console;

use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class Quote extends Command
{
    protected function configure()
    {
        $this->setName('quote:create');
        $this->setDescription('Demo command line');

        parent::configure();
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {

        $objectManager = \Magento\Framework\App\ObjectManager::getInstance();

        $storeManager = $objectManager->get('\Magento\Store\Model\StoreManagerInterface');
        $store = $storeManager->getStore();
        $websiteId = $storeManager->getStore()->getWebsiteId();

        $firstName = 'John';
        $lastName = 'Doe';
        $email = 'doe.r@';
        $password = 'Test@123';

        $address = [
            'customer_address_id' => '',
            'prefix' => '',
            'firstname' => $firstName,
            'middlename' => '',
            'lastname' => $lastName,
            'suffix' => '',
            'company' => '',
            'street' => [
                '0' => 'Customer Address 1', // this is mandatory
                '1' => 'Customer Address 2' // this is optional
            ],
            'city' => 'New York',
            'country_id' => 'US', // two letters country code
            'region' => 'New York', // can be empty '' if no region
            'region_id' => '43', // can be empty '' if no region_id
            'postcode' => '10450',
            'telephone' => '123-456-7890',
            'fax' => '',
            'save_in_address_book' => 1,
        ];

        $customerFactory = $objectManager->get('\Magento\Customer\Model\CustomerFactory')->create();

        /**
         * check whether the email address is already registered or not
         */
        $customer = $customerFactory->setWebsiteId($websiteId)->loadByEmail($email);
        if (!$customer->getId()) {
            try {
                $customer = $objectManager->get('\Magento\Customer\Model\CustomerFactory')->create();
                $customer->setWebsiteId($websiteId);
                $customer->setEmail($email);
                $customer->setFirstname($firstName);
                $customer->setLastname($lastName);
                $customer->setPassword($password);
                $customer->save();

                $customer->setConfirmation(null);
                $customer->save();

                $customAddress = $objectManager->get('\Magento\Customer\Model\AddressFactory')->create();
                $customAddress->setData($address)
                    ->setCustomerId($customer->getId())
                    ->setIsDefaultBilling('1')
                    ->setIsDefaultShipping('1')
                    ->setSaveInAddressBook('1');
                $customAddress->save();
            } catch (Exception $e) {
                echo $e->getMessage();
            }
        }

        $customer = $objectManager->get('\Magento\Customer\Api\CustomerRepositoryInterface')->getById($customer->getId());
        try {

            $quoteFactory = $objectManager->get('\Magento\Quote\Model\QuoteFactory')->create();
            $quoteFactory->setStore($store);
            $quoteFactory->setCurrency();
            $quoteFactory->assignCustomer($customer);

            $productIds = [337 => 2, 338 => 3];
            foreach ($productIds as $productId => $qty) {
                $product = $objectManager->get('\Magento\Catalog\Model\ProductRepository')->getById($productId);// get product by product id
                $quoteFactory->addProduct($product, $qty);  // add products to quote
            }

            /*
             * Set Address to quote
             */
            $quoteFactory->getBillingAddress()->addData($address);
            $quoteFactory->getShippingAddress()->addData($address);

            /*
             * Collect Rates and Set Shipping & Payment Method
             */
            $shippingAddress = $quoteFactory->getShippingAddress();
            $shippingAddress->setCollectShippingRates(true)
                ->collectShippingRates()
                ->setShippingMethod('flatrate_flatrate'); //shipping method

            $quoteFactory->setPaymentMethod('checkmo'); //payment method
            $quoteFactory->setInventoryProcessed(false);
            $quoteFactory->save();

            /*
             * Set Sales Order Payment
             */
            $quoteFactory->getPayment()->importData(['method' => 'checkmo']);

            /*
             * Collect Totals & Save Quote
             */
            $quoteFactory->collectTotals()->save();

            /*
             * Create Order From Quote
             */
            $order = $objectManager->get('\Magento\Quote\Model\QuoteManagement')->submit($quoteFactory);
            $order->setEmailSent(0);
            $output->writeln('Order Id:' . $order->getRealOrderId());
        } catch (Exception $e) {
            $output->writeln($e->getMessage());
        }
        try {

            $quoteFactory = $objectManager->get('\Magento\Quote\Model\QuoteFactory')->create();
            $quoteFactory->setStore($store);
            $quoteFactory->setCurrency();
            $quoteFactory->assignCustomer($customer);

            $productIds = [337 => 2, 338 => 3];
            foreach ($productIds as $productId => $qty) {
                $product = $objectManager->get('\Magento\Catalog\Model\ProductRepository')->getById($productId);// get product by product id
                $quoteFactory->addProduct($product, $qty);  // add products to quote
            }

            /*
             * Set Address to quote
             */
            $quoteFactory->getBillingAddress()->addData($address);
            $quoteFactory->getShippingAddress()->addData($address);

            /*
             * Collect Rates and Set Shipping & Payment Method
             */
            $shippingAddress = $quoteFactory->getShippingAddress();
            $shippingAddress->setCollectShippingRates(true)
                ->collectShippingRates()
                ->setShippingMethod('flatrate_flatrate'); //shipping method

            $quoteFactory->setPaymentMethod('checkmo'); //payment method
            $quoteFactory->setInventoryProcessed(false);
            $quoteFactory->save();

            /*
             * Set Sales Order Payment
             */
            $quoteFactory->getPayment()->importData(['method' => 'checkmo']);

            /*
             * Collect Totals & Save Quote
             */
            $quoteFactory->collectTotals()->save();

            /*
             * Create Order From Quote
             */
            $order = $objectManager->get('\Magento\Quote\Model\QuoteManagement')->submit($quoteFactory);
            $order->setEmailSent(0);
            $output->writeln('Order Id:' . $order->getRealOrderId());
        } catch (Exception $e) {
            $output->writeln($e->getMessage());
        }
    }
}

```
**Result:**
![image](https://user-images.githubusercontent.com/51680850/70894405-9e35dc80-1ff5-11ea-8ffa-3718428b2b87.png)

**Comment by engcom-Charlie** [comment]
@stephan-cream i skipped creation of new customer and just changed customer's email in the database and just load customer
![image](https://user-images.githubusercontent.com/51680850/70897690-caecf280-1ffb-11ea-9ddb-08db82a4e8c9.png)
```
<?php

namespace Module\Vendor\Console;

use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class Quote extends Command
{
    protected function configure()
    {
        $this->setName('quote:create');
        $this->setDescription('Demo command line');

        parent::configure();
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {

        $objectManager = \Magento\Framework\App\ObjectManager::getInstance();
        $state = $objectManager->get('\Magento\Framework\App\State');
        $state->setAreaCode(\Magento\Framework\App\Area::AREA_GLOBAL);
        $storeManager = $objectManager->get('\Magento\Store\Model\StoreManagerInterface');
        $store = $storeManager->getStore();
        $websiteId = 1;

        $firstName = 'John';
        $lastName = 'Doe';
        $address = [
            'customer_address_id' => '',
            'prefix' => '',
            'firstname' => $firstName,
            'middlename' => '',
            'lastname' => $lastName,
            'suffix' => '',
            'company' => '',
            'street' => [
                '0' => 'Customer Address 1', // this is mandatory
                '1' => 'Customer Address 2' // this is optional
            ],
            'city' => 'New York',
            'country_id' => 'US', // two letters country code
            'region' => 'New York', // can be empty '' if no region
            'region_id' => '43', // can be empty '' if no region_id
            'postcode' => '10450',
            'telephone' => '123-456-7890',
            'fax' => '',
            'save_in_address_book' => 1,
        ];

        $customerFactory = $objectManager->get('\Magento\Customer\Model\CustomerFactory')->create();
        $customer = $objectManager->get('\Magento\Customer\Api\CustomerRepositoryInterface')->getById(1);
        try {

            $quoteFactory = $objectManager->get('\Magento\Quote\Model\QuoteFactory')->create();
            $quoteFactory->setStore($store);
            $quoteFactory->setCurrency();
            $quoteFactory->assignCustomer($customer);

            $productIds = [1 => 1];
            foreach ($productIds as $productId => $qty) {
                $product = $objectManager->get('\Magento\Catalog\Model\ProductRepository')->getById($productId);// get product by product id
                $quoteFactory->addProduct($product, $qty);  // add products to quote
            }

            /*
             * Set Address to quote
             */
            $quoteFactory->getBillingAddress()->addData($address);
            $quoteFactory->getShippingAddress()->addData($address);

            /*
             * Collect Rates and Set Shipping & Payment Method
             */
            $shippingAddress = $quoteFactory->getShippingAddress();
            $shippingAddress->setCollectShippingRates(true)
                ->collectShippingRates()
                ->setShippingMethod('flatrate_flatrate'); //shipping method

            $quoteFactory->setPaymentMethod('checkmo'); //payment method
            $quoteFactory->setInventoryProcessed(false);
            $quoteFactory->save();

            /*
             * Set Sales Order Payment
             */
            $quoteFactory->getPayment()->importData(['method' => 'checkmo']);

            /*
             * Collect Totals & Save Quote
             */
            $quoteFactory->collectTotals()->save();

            /*
             * Create Order From Quote
             */
            $order = $objectManager->get('\Magento\Quote\Model\QuoteManagement')->submit($quoteFactory);
            $order->setEmailSent(0);
            $output->writeln('Order Id:' . $order->getRealOrderId());
        } catch (Exception $e) {
            $output->writeln($e->getMessage());
        }
        try {

            $quoteFactory = $objectManager->get('\Magento\Quote\Model\QuoteFactory')->create();
            $quoteFactory->setStore($store);
            $quoteFactory->setCurrency();
            $quoteFactory->assignCustomer($customer);

            $productIds = [1 => 1];
            foreach ($productIds as $productId => $qty) {
                $product = $objectManager->get('\Magento\Catalog\Model\ProductRepository')->getById($productId);// get product by product id
                $quoteFactory->addProduct($product, $qty);  // add products to quote
            }

            /*
             * Set Address to quote
             */
            $quoteFactory->getBillingAddress()->addData($address);
            $quoteFactory->getShippingAddress()->addData($address);

            /*
             * Collect Rates and Set Shipping & Payment Method
             */
            $shippingAddress = $quoteFactory->getShippingAddress();
            $shippingAddress->setCollectShippingRates(true)
                ->collectShippingRates()
                ->setShippingMethod('flatrate_flatrate'); //shipping method

            $quoteFactory->setPaymentMethod('checkmo'); //payment method
            $quoteFactory->setInventoryProcessed(false);
            $quoteFactory->save();

            /*
             * Set Sales Order Payment
             */
            $quoteFactory->getPayment()->importData(['method' => 'checkmo']);

            /*
             * Collect Totals & Save Quote
             */
            $quoteFactory->collectTotals()->save();

            /*
             * Create Order From Quote
             */
            $order = $objectManager->get('\Magento\Quote\Model\QuoteManagement')->submit($quoteFactory);
            $order->setEmailSent(0);
            $output->writeln('Order Id:' . $order->getRealOrderId());
        } catch (Exception $e) {
            $output->writeln($e->getMessage());
        }
    }
}

```

**Comment by stephan-cream** [comment]
It is fine if it is fixed in 2.4.0 but we have no idea when this will be released.
My bug report was not only placed on 2017 but also refers to Magento versions up to 2.3.2.
If there has been a fix released for this but it won't be released for current released versions, we would at least like to know what that fix is so we can implement a patch for this until 2.4.0 is released.

Furthermore your test differs from my description so i can't be sure if it is actually fixed.
**See my test (this has to be tested):**
$quote = $this->quoteFactory->create(); // or you can do this with an existing one
...
try { $this->quoteRepository->save($quote); }
catch () { $this->quoteRepository->save($quote); // 2nd save for the same quote }

**Your test (which is different):**
$quote = $this->quoteFactory->create();
try { $quote->collectTotals()->save();  }
catch { // log error}

$quote = $this->quoteFactory->create();
try { $quote->collectTotals()->save(); }
catch { // log error }

**Comment by engcom-Charlie** [comment]
@stephan-cream 
Is your `quoteRepository` using `CartRepositoryInterface`?

I tried again with your new update and result is same as before:
![image](https://user-images.githubusercontent.com/51680850/70900532-b6abf400-2001-11ea-895e-2479fc6f6738.png)

```
<?php

namespace Mod\HelloWorld\Console;

use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class Sayhello extends Command
{
    protected function configure()
    {
        $this->setName('quote:create');
        $this->setDescription('Demo command line');

        parent::configure();
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {

        $objectManager = \Magento\Framework\App\ObjectManager::getInstance();
        $state = $objectManager->get('\Magento\Framework\App\State');
        $state->setAreaCode(\Magento\Framework\App\Area::AREA_GLOBAL);
        $storeManager = $objectManager->get('\Magento\Store\Model\StoreManagerInterface');
        $store = $storeManager->getStore();
        $websiteId = 1;

        $firstName = 'John';
        $lastName = 'Doe';
        $address = [
            'customer_address_id' => '',
            'prefix' => '',
            'firstname' => $firstName,
            'middlename' => '',
            'lastname' => $lastName,
            'suffix' => '',
            'company' => '',
            'street' => [
                '0' => 'Customer Address 1', // this is mandatory
                '1' => 'Customer Address 2' // this is optional
            ],
            'city' => 'New York',
            'country_id' => 'US', // two letters country code
            'region' => 'New York', // can be empty '' if no region
            'region_id' => '43', // can be empty '' if no region_id
            'postcode' => '10450',
            'telephone' => '123-456-7890',
            'fax' => '',
            'save_in_address_book' => 1,
        ];

        $customerFactory = $objectManager->get('\Magento\Customer\Model\CustomerFactory')->create();
        $customer = $objectManager->get('\Magento\Customer\Api\CustomerRepositoryInterface')->getById(1);
        try {

            $quoteFactory = $objectManager->get('\Magento\Quote\Model\QuoteFactory')->create();
            $quoteFactory->setStore($store);
            $quoteFactory->setCurrency();
            $quoteFactory->assignCustomer($customer);

            $productIds = [1 => 1];
            foreach ($productIds as $productId => $qty) {
                $product = $objectManager->get('\Magento\Catalog\Model\ProductRepository')->getById($productId);// get product by product id
                $quoteFactory->addProduct($product, $qty);  // add products to quote
            }

            /*
             * Set Address to quote
             */
            $quoteFactory->getBillingAddress()->addData($address);
            $quoteFactory->getShippingAddress()->addData($address);

            /*
             * Collect Rates and Set Shipping & Payment Method
             */
            $shippingAddress = $quoteFactory->getShippingAddress();
            $shippingAddress->setCollectShippingRates(true)
                ->collectShippingRates()
                ->setShippingMethod('flatrate_flatrate'); //shipping method

            $quoteFactory->setPaymentMethod('checkmo'); //payment method
            $quoteFactory->setInventoryProcessed(false);
            $quoteFactory->save();

            /*
             * Set Sales Order Payment
             */
            $quoteFactory->getPayment()->importData(['method' => 'checkmo']);

            /*
             * Collect Totals & Save Quote
             */
            $quoteFactory->collectTotals()->save();

            /*
             * Create Order From Quote
             */
            $order = $objectManager->get('\Magento\Quote\Model\QuoteManagement')->submit($quoteFactory);
            $order->setEmailSent(0);
            $output->writeln('Order Id:' . $order->getRealOrderId());
        } catch (Exception $e) {
            $quoteFactory->collectTotals()->save();
        }
    }
}

```

**Comment by stephan-cream** [comment]
@engcom-Charlie
Apparently it no longer is an issue in 2.4.0.
But as i said before:

`It is fine if it is fixed in 2.4.0 but we have no idea when this will be released.
My bug report was not only placed on 2017 but also refers to Magento versions up to 2.3.2.
If there has been a fix released for this but it won't be released for current released versions, we would at least like to know what that fix is so we can implement a patch for this until 2.4.0 is released.`

**Comment by exequielserfe** [comment]
We faced this problem in Magento 2.3.5-p2, trying to create a new order from the backend.

We fixed it creating a custom patch to the **magento/framework** module _Magento\Framework\Model\AbstractModel::validateBeforeSave()_ with this code:

```
/**
     * Validate model before saving it
     *
     * @return $this
     * @throws \Magento\Framework\Validator\Exception
     */
    public function validateBeforeSave()
    {
        $validator = $this->_getValidatorBeforeSave();
        if ($validator && !$validator->isValid($this)) {
            $errors = $validator->getMessages();
            // Handle associative array messages. Fix issue with backend order email validation
            $errors = array_values($errors);
            foreach ($errors as $key => $value) {
              if (is_array($value)) {
                  // Prevent duplicate messages
                  $value = array_unique($value);
                  $errors[$key] = implode(PHP_EOL, $value);
              }
            }
            // End error message array customization
            $exception = new \Magento\Framework\Validator\Exception(
                new Phrase(implode(PHP_EOL, $errors))
            );
            foreach ($errors as $errorMessage) {
                $exception->addMessage(new \Magento\Framework\Message\Error($errorMessage));
            }
            throw $exception;
        }
        return $this;
    }
```

Hope this helps somebody else

**Comment by dverkade** [comment]
Adding a message here, so that the issue is no longer stale and will be auto closed. Closing confirmed issues is not the way a project should be handled and it should be open until it's resolved.

**Comment by dverkade** [comment]
Adding a message here, so that the issue is no longer stale and will be auto closed. Closing confirmed issues is not the way a project should be handled and it should be open until it's resolved.

**Comment by engcom-Hotel** [comment]
Hi @stephan-cream and @exequielserfe 

We have tried to reproduce this issue in Magento v 2.4-develop but it is not reproducible. This issue has already been fixed in previous releases.

So I am closing this issue for now. If you again able to reproduce the issue in 2.4-develop version than you can reopen this ticket or create new ticket.

Thanks
